### PR TITLE
Added explicit specifier to ctors in core

### PR DESCRIPTION
**Pull request by msandre**
This PR replaces #2527 on a new branch. The new branch avoids merge conflicts and has a cleaner commit history. It compiles fine on my machine now.

closes #2527

### CODE REVIEW

**Comment by philbucher** [comment]
@msandre I think you are not porting all the apps
I suggest compiling the ones that are compiled in nightly

**Comment by msandre** [comment]
@KratosMultiphysics/technical-committee this is ready for review.

The changes mainly affected points and geometries. For example, there exist functions like `void f1(const Point&)` which are being called like `f1(node.Coordinates())` or `void f2(const Line2D2<Node<3>>&)` called like `f2(geom)` (not as often). Where it was straightforward to do so, I changed the parameter being passed to the function to eliminate the temporary. But I did not change function declarations so there are still temporaries in those places where the function requires overly restrictive parameters.

**Comment by pooyan-dadvand** [comment]
Now I see why you closed the other one.

**Comment by pooyan-dadvand** [comment]
Seems that the change of tests folder is causing conflicts....

**Comment by pooyan-dadvand** [comment]
Forwarding to @KratosMultiphysics/implementation-committee

**Comment by loumalouomega** [comment]
I solved the conflicts. I will add to ours pending stuff

**Comment by msandre** [comment]
I thought it was passing. Thx for fixing @loumalouomega

**Comment by loumalouomega** [comment]
imagen

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp:160]: would it work here? (i took out the "Coordinates()")

     const auto local_point_decomp = PointType(integration_points_slave[point_number])

Comment (msandre) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp:160]: Hi Riccardo, I checked it and yes it works. This could be cleaned up more with `const auto& local_point_decomp = integration_points_slave[point_number];`, but aim of this PR was to add the explicit and replace broken implicit conversions with explicit conversions (essentially not changing the behavior/performance). I could go ahead and clean these up if you want, but I would prefer the application developers do this (ideally in subsequent PRs) since I'm not familiar with this app (or many other apps). Also there is code repetition which I think is being generated by some automated tool. It's a bit of effort to fix all of these by hand.

**Comment by RiccardoRossi** [comment]
to me this is ok and i would approve it as it is, however i would like @KratosMultiphysics/technical-committee to take a look

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp
Comment (RiccardoRossi) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp:160]: ok

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand @roigcarlo @jcotela we asked for this PR and i would approve it unless you have some important concern about it...

**Comment by pooyan-dadvand** [comment]
I agree. Only one question: Does it compatible with AMatrix?

**Comment by GuillermoCasas** [comment]
@msandre, @loumalouomega speaking here: the file in conflict is a not existing file currently, please remove it. The new implementation of this file is in base_contact_search.cpp

**Review by GuillermoCasas — CHANGES_REQUESTED** [review]
Refer to my comment

**Comment by adityaghantasala** [comment]
@msandre the concern about compatibility with Amatrix is to be checked other wise @KratosMultiphysics/implementation-committee  has no objections for this PR.

**Comment by msandre** [comment]
@GuillermoCasas , @loumalouomega I removed the files. @adityaghantasala I tested with `AMATRIX_DIR` active and don't see any conflicts with this PR. There were some existing conflicts in some of the applications which I had to turn off to compile.

Please let me know if there are any more changes to be made.

**Comment by loumalouomega** [comment]
There are some Travis errors:

~~~
In file included from /home/travis/build/KratosMultiphysics/Kratos/cmake_build/kratos/cotire/KratosCore_CXX_unity_87_172.cxx:13:0:
/home/travis/build/KratosMultiphysics/Kratos/kratos/tests/cpp_tests/geometries/test_line_2d_2.cpp: In member function ‘virtual void Kratos::Testing::TestLine2D2PointLocalCoordinates::TestFunction()’:
/home/travis/build/KratosMultiphysics/Kratos/kratos/tests/cpp_tests/geometries/test_line_2d_2.cpp:126:36: error: conversion from ‘boost::numeric::ublas::vector_binary_traits<Kratos::array_1d<double, 3ul>, Kratos::array_1d<double, 3ul>, boost::numeric::ublas::scalar_plus<double, double> >::result_type {aka boost::numeric::ublas::vector_binary<Kratos::array_1d<double, 3ul>, Kratos::array_1d<double, 3ul>, boost::numeric::ublas::scalar_plus<double, double> >}’ to non-scalar type ‘Kratos::Point’ requested
         Point centre = geom_pts[0] + geom_pts[1];
                                     
In file included from /home/travis/build/KratosMultiphysics/Kratos/cmake_build/kratos/cotire/KratosCore_CXX_unity_87_172.cxx:14:0:
/home/travis/build/KratosMultiphysics/Kratos/kratos/tests/cpp_tests/geometries/test_line_3d_2.cpp: In member function ‘virtual void Kratos::Testing::TestLine3D2PointLocalCoordinates::TestFunction()’:
/home/travis/build/KratosMultiphysics/Kratos/kratos/tests/cpp_tests/geometries/test_line_3d_2.cpp:125:36: error: conversion from ‘boost::numeric::ublas::vector_binary_traits<Kratos::array_1d<double, 3ul>, Kratos::array_1d<double, 3ul>, boost::numeric::ublas::scalar_plus<double, double> >::result_type {aka boost::numeric::ublas::vector_binary<Kratos::array_1d<double, 3ul>, Kratos::array_1d<double, 3ul>, boost::numeric::ublas::scalar_plus<double, double> >}’ to non-scalar type ‘Kratos::Point’ requested
         Point centre = geom_pts[0] + geom_pts[1];
                                     
~~~

**Comment by msandre** [comment]
ping @GuillermoCasas , @loumalouomega. I addressed your comment.

**Comment by GuillermoCasas** [comment]
The @KratosMultiphysics/implementation-committee supports this change. passing it on to the @KratosMultiphysics/technical-committee

File: kratos/utilities/variable_utils.cpp
Comment (philbucher) [inline kratos/utilities/variable_utils.cpp:506]: is this needed in the scope of this PR?

**Review by GuillermoCasas — APPROVED** [review]

**Review by GuillermoCasas — DISMISSED** [review]
Sorry, I approved by mistake. The technical committee should decide.

**Comment by msandre** [comment]
@KratosMultiphysics/technical-committee please let me know when you are ready to approve. Then I will fix the remaining conflicts since they keep popping up when master changes.

**Comment by pooyan-dadvand** [comment]
Thanks @msandre for your effort on this. 

@KratosMultiphysics/technical-committee approves this PR.

**Comment by philbucher** [comment]
@msandre I would recommend to wait until the release-brach is created (I think next Friday) to avoid having errors in the release

I would approve right after that (since @KratosMultiphysics/technical-committee approves) such that you don’t have to continuously solve conflicts 

Is that ok for you?

**Comment by jcotela** [comment]
@philbucher we are trying to solve the conflicts now, so that this PR can go into the release.

**Comment by philbucher** [comment]
Ok cool :)

**Review by jcotela — APPROVED** [review]
:+1: approving. Looks like remaining conflicts have been solved. Thanks for the effort!

**Comment by msandre** [comment]
Thanks :+1:

**Comment by jcotela** [comment]
Added the API breaker label for future reference, to indicate that this can make working code to stop compiling.

@KratosMultiphysics/altair you may want to check your applications.

**Pull request merged**
